### DISCUSSION TITLE
Can external locking be used make the state machine truly thread-safe?

### DISCUSSION THREAD

**Discussion by jmo-gelsight** [dotnet-state-machine/stateless#596] [General]
I understand that Stateless is not designed to be used from multiple threads.  But even after reading the [lengthy discussion about the issue here](https://github.com/dotnet-state-machine/stateless/issues/527), I am still unclear on one point:

If I use the `StateMachine` simultaneously from multiple threads with `FiringMode.Queued` but I protect every single call to the `Fire()` function with a simple lock like so...

```
lock(_myLock)
    Machine.Fire(Trigger.DoSomething);
```
    
Does _that_ make my code thread safe?  Would any problems still remain?  Because frankly, that's what I am doing and it seems to work perfectly.... _almost_ all the time.  And I am unable to determine if that "almost" is due to other bugs of mine or the fact that what I'm doing is still not safe.

**Comment by crozone** [comment]
Yep, if you funnel every interaction with the state machine through a lock, that will effectively serialize all operations and be thread-safe.

Just note that this only works for the synchronous `Fire()` call. It won't behave correctly if the state machine contains async handlers. However, if you keep everything synchronous, this should work just fine.

> And I am unable to determine if that "almost" is due to other bugs of mine or the fact that what I'm doing is still not safe.

Are other threads touching any part of the state machine *outside* of the lock? Because you can still get weird behaviour if you're interacting with memory that the state machine is also interacting with, unless you also put a lock around that.

**Comment by jmoleary** [comment]
No.  No other threads are doing that.  After startup, configuring the states, and calling `Activate`, the only `StateMachine` function I call is `Fire()`.  So whatever the bug is, I guess it is somewhere else in my code.

But since I was unsure about the answer I would get here, yesterday I actually went ahead and created a class derived from `StateMachine` that turns every every single "Fire" call into an `Action` object and then funnels it through a .NET [Channel ](https://learn.microsoft.com/en-us/dotnet/api/system.threading.channels.channel-1?view=net-8.0&devlangs=csharp&f1url=%3FappId%3DDev17IDEF1%26l%3DEN-US%26k%3Dk(System.Threading.Channels.Channel%25601)%3Bk(SolutionItemsProject)%3Bk(SolutionItemsProject)%3Bk(DevLang-csharp)%26rd%3Dtrue)object so that all calls would be handled... well... serially.  My kludgey, poor-man's version of   "gao-artur"'s `FiringMode.Serial`  (Wouldn't work for `FireAsync` either though).  

From what you are saying it sounds like I don't need it.  As long as I stick with `Fire` lock every time, and avoid `FireAsync`  I'll make my class  just hide every base class version and do that.  Maybe throw `NotImplementedException` for `FireAsync` just to be safe.  Thanks!
